tests/examples/token/test_vipercoin.py

```diff
@@ -0,0 +1,177 @@
+
+    # Negative transfer value.
+    assert_tx_failed(
+        tester=token_tester,
+        function_to_test=lambda: tester.c.transfer(token_tester.accounts[1], -1),
+        exception=ValueOutOfBounds
+    )
+
+
+def test_approve_allowance(token_tester):
+    assert token_tester.c.allowance(token_tester.accounts[0], token_tester.accounts[1]) == 0
+    assert token_tester.c.approve(token_tester.accounts[1], 10) is True
+    assert token_tester.c.allowance(token_tester.accounts[0], token_tester.accounts[1]) == 10
+
+
+def test_transferFrom(token_tester, assert_tx_failed):
+    a0 = token_tester.accounts[0]
+    a1 = token_tester.accounts[1]
+    k1 = token_tester.k1
+    a2 = token_tester.accounts[2]
+    k2 = token_tester.k2
+    contract = token_tester.c
+
+    # Allow 10 token transfers. Account a1 is allowed to spend 10 tokens of a0's account.
+    ALLOWANCE = 10
+    assert contract.approve(a1, 10) is True
+    assert contract.allowance(a0, a1) == ALLOWANCE
+
+    assert contract.transferFrom(a0, a2, 3, sender=k1) is True  # a1 may transfer.
+    assert contract.balanceOf(a0) == TOKEN_TOTAL_SUPPLY - 3
+    assert contract.balanceOf(a1) == 0
+    assert contract.balanceOf(a2) == 3
+    assert contract.allowance(a0, a1) == ALLOWANCE - 3
+
+    # a2 may not transfer.
+    assert contract.transferFrom(a0, a2, ALLOWANCE, sender=k2) is False
+
+    # Negative transfer value.
+    assert_tx_failed(
+        tester=token_tester,
+        function_to_test=lambda: contract.transferFrom(a0, a2, -1, sender=k1),
+        exception=ValueOutOfBounds
+    )
+
+    # Transfer more than allowance:
+    assert contract.transferFrom(a0, a2, 8, sender=k1) is False
+    assert contract.balanceOf(a0) == TOKEN_TOTAL_SUPPLY - 3
+    assert contract.balanceOf(a1) == 0
+    assert contract.balanceOf(a2) == 3
+    assert contract.allowance(a0, a1) == ALLOWANCE - 3
+
+    # Transfer exact amount left in allowance:
+    allowance_left = contract.allowance(a0, a1)
+    assert contract.transferFrom(a0, a2, allowance_left, sender=k1) is True
+    assert contract.balanceOf(a0) == TOKEN_TOTAL_SUPPLY - ALLOWANCE
+    assert contract.balanceOf(a1) == 0
+    assert contract.balanceOf(a2) == ALLOWANCE
+    assert contract.allowance(a0, a1) == 0
+
+
+def test_transfer_event(token_tester):
+    a0 = token_tester.accounts[0]
+    a1 = token_tester.accounts[1]
+    k1 = token_tester.k1
+    a2 = token_tester.accounts[2]
+
+    contract = token_tester.c
+    chain = token_tester.s
+
+    assert contract.transfer(a1, 1) is True
+
+    logs = chain.head_state.receipts[-1].logs[-1]
+    event_id = utils.bytes_to_int(utils.sha3(bytes('Transfer(address,address,uint256)', 'utf-8')))
+
+    assert logs.topics[0] == event_id
+    assert contract.translator.event_data[event_id] == {
+        'types': ['address', 'address', 'uint256'],
+        'name': 'Transfer', 'names': ['_from', '_to', '_value'],
+        'indexed': [True, True, False], 'anonymous': False
+    }
```

Comment:
@DavidKnott this seems crazy to me, it would be nice to have some event helpers in the viper module to decode all of this nonsense more easily.

Reply:
@fubuloubu If I'm understanding you correctly, you're saying you'd like to have test helper to test logs in a clearer way (hiding the specific checks within the helper).

Reply:
Yeah, I guess Web3 handles that for any production uses, a test helper here would be adequate.